File: wear/src/main/java/com/pdyjak/powerampwear/MusicLibraryNavigatorImpl.java
package com.pdyjak.powerampwear;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.pdyjak.powerampwear.music_browser.MusicLibraryNavigator;
import com.pdyjak.powerampwear.music_browser.albums.AlbumItem;
import com.pdyjak.powerampwear.music_browser.artists.ArtistItem;
import com.pdyjak.powerampwear.music_browser.categories.CategoryItem;
import com.pdyjak.powerampwear.music_browser.files.FileItem;
import com.pdyjak.powerampwear.music_browser.folders.FolderItem;

import java.util.HashSet;
import java.util.Set;

public class MusicLibraryNavigatorImpl implements MusicLibraryNavigator {

    @NonNull
    private final Set<MusicLibraryNavigator.Listener> mMusicBrowserListeners = new HashSet<>();

    @Override
    public void selectCategory(@NonNull CategoryItem item, boolean fromPlayer) {
        Set<Listener> copy = new HashSet<>(mMusicBrowserListeners);
        for (MusicLibraryNavigator.Listener listener : copy) {
            listener.onCategorySelected(item, fromPlayer);
        }
    }

    @Override
    public void selectFolder(@NonNull FolderItem item, boolean fromPlayer,
                             @Nullable String scrollTo) {
        Set<MusicLibraryNavigator.Listener> copy = new HashSet<>(mMusicBrowserListeners);
        for (MusicLibraryNavigator.Listener listener : copy) {
            listener.onFolderSelected(item, fromPlayer, scrollTo);
        }
    }

    @Override
    public void selectAlbum(@NonNull AlbumItem item, boolean fromPlayer) {
        Set<MusicLibraryNavigator.Listener> copy = new HashSet<>(mMusicBrowserListeners);
        for (MusicLibraryNavigator.Listener listener : copy) {
            listener.onAlbumSelected(item, fromPlayer);
        }
    }

    @Override
    public void selectFile(@NonNull FileItem item, boolean fromPlayer) {
        Set<MusicLibraryNavigator.Listener> copy = new HashSet<>(mMusicBrowserListeners);
        for (MusicLibraryNavigator.Listener listener : copy) {
            listener.onFileSelected(item, fromPlayer);
        }
    }

    @Override
    public void selectArtist(@NonNull ArtistItem item, boolean fromPlayer) {
        Set<MusicLibraryNavigator.Listener> copy = new HashSet<>(mMusicBrowserListeners);
        for (MusicLibraryNavigator.Listener listener : copy) {
            listener.onArtistSelected(item, fromPlayer);
        }
    }

    @Override
    public void addLibraryNavigationListener(@NonNull Listener listener) {
        mMusicBrowserListeners.add(listener);
    }

    @Override
    public void removeLibraryNavigationListener(@NonNull Listener listener) {
        mMusicBrowserListeners.remove(listener);
    }
}

File: wear/src/main/java/com/pdyjak/powerampwear/MusicLibraryNavigatorImpl.kt
package com.pdyjak.powerampwear

import com.pdyjak.powerampwear.common.Event
import com.pdyjak.powerampwear.music_browser.AlbumSelectedEventArgs
import com.pdyjak.powerampwear.music_browser.ArtistSelectedEventArgs
import com.pdyjak.powerampwear.music_browser.CategorySelectedEventArgs
import com.pdyjak.powerampwear.music_browser.FileSelectedEventArgs
import com.pdyjak.powerampwear.music_browser.FolderSelectedEventArgs
import com.pdyjak.powerampwear.music_browser.MusicLibraryNavigator
import com.pdyjak.powerampwear.music_browser.albums.AlbumItem
import com.pdyjak.powerampwear.music_browser.artists.ArtistItem
import com.pdyjak.powerampwear.music_browser.categories.CategoryItem
import com.pdyjak.powerampwear.music_browser.files.FileItem
import com.pdyjak.powerampwear.music_browser.folders.FolderItem

class MusicLibraryNavigatorImpl : MusicLibraryNavigator {
    override val onCategorySelected: Event<CategorySelectedEventArgs> = Event()
    override val onFolderSelected: Event<FolderSelectedEventArgs> = Event()
    override val onAlbumSelected: Event<AlbumSelectedEventArgs> = Event()
    override val onArtistSelected: Event<ArtistSelectedEventArgs> = Event()
    override val onFileSelected: Event<FileSelectedEventArgs> = Event()

    override fun selectCategory(item: CategoryItem, fromPlayer: Boolean, scrollTo: String?) {
        onCategorySelected(CategorySelectedEventArgs(item, fromPlayer, scrollTo))
    }

    override fun selectFolder(item: FolderItem, fromPlayer: Boolean, scrollTo: String?) {
        onFolderSelected(FolderSelectedEventArgs(item, fromPlayer, scrollTo))
    }

    override fun selectAlbum(item: AlbumItem, fromPlayer: Boolean, scrollTo: String?) {
        onAlbumSelected(AlbumSelectedEventArgs(item, fromPlayer, scrollTo))
    }

    override fun selectFile(item: FileItem, fromPlayer: Boolean) {
        onFileSelected(FileSelectedEventArgs(item, fromPlayer))
    }

    override fun selectArtist(item: ArtistItem, fromPlayer: Boolean) {
        onArtistSelected(ArtistSelectedEventArgs(item, fromPlayer))
    }
}
